internal/service/user_service.go

```diff
@@ -0,0 +1,145 @@
+package service
+
+import (
+	"context"
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gorm.io/gorm"
+
+	"github.com/smailic05/TaskInfoblox/internal/model"
+	"github.com/smailic05/TaskInfoblox/internal/pb"
+)
+
+const (
+	ErrUserNotExist = "The user does not exist"
+	ErrUserExist    = "Error, user already exists"
+	Success         = "Success"
+	Deleted         = "Deleted"
+)
+
+type Repository interface {
+	Load(user model.User) ([]model.User, error)
+	LoadOne(user model.User) (model.User, error)
+	Store(user model.User) error
+	Update(user model.User) error
+	DeleteUser(user model.User) error
+}
+
+type UserService struct {
+	pb.UnimplementedUserServiceServer
+	repo Repository
+}
+
+func New(repo Repository) *UserService {
+	return &UserService{repo: repo}
+}
+
+func (s *UserService) AddUser(ctx context.Context, addUser *pb.AddUserRequest) (*pb.AddUserResponse, error) {
+	user := model.User{
+		Address:  addUser.Address,
+		Username: addUser.Username,
+		Phone:    addUser.Phone}
+	user, err := s.repo.LoadOne(user)
+	if err != gorm.ErrRecordNotFound {
+		return nil, status.Errorf(codes.AlreadyExists, "%s: when adding user", ErrUserExist)
+	}
+	err = s.repo.Store(user)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.AddUserResponse{Response: Success}, nil
+}
+
+func (s *UserService) DeleteUser(ctx context.Context, deleteUser *pb.DeleteUserRequest) (*pb.DeleteUserResponse, error) {
+	user := model.User{
+		Address:  deleteUser.Address,
+		Username: deleteUser.Username,
+		Phone:    deleteUser.Phone}
+	err := s.repo.DeleteUser(user)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.DeleteUserResponse{}, nil
+}
+
+func (s *UserService) FindUser(findUser *pb.FindUserRequest, srv pb.UserService_FindUserServer) error {
```

Comment:
Is HTTP 1.2 able to handle streaming? I'm not sure it is.

Reply:
Doesn't grpc support http/2?

Reply:
Yes, it does, but I mean you have grpc to http gateway here. So you're allowing both grpc and http requests on 2 different ports.

Reply:
So http 1.2 does not support streaming, but as mentioned in the grpc gateway documentation, the gateway transform the streaming to newline delimited JSON streams, so I think it works correctly

Reply:
Got it, thanks